File: python/demoapp_smoke.py
"""Full-demo build smoke: runs the Python demoapp's build path against a
REAL Go display host (driven by interop_test.go), the Python mirror of Go's
TestDemoBuildsOverService.

It builds the whole nine-tab main window, the protocol companion window,
the About dialog, an MDI document + dock entry, exercises the tab/window
properties, and fires every desktop-action app verb - proving the display
session accepts the full demo vocabulary from a Python client. Shell-bearing
scripts (terminal windows, secondary apps) are omitted, as in the Go test.

Prints OK on success.
"""

import os
import sys

sys.path.insert(0, os.path.dirname(__file__))

import kittytk  # noqa: E402
from demoapp import scripts  # noqa: E402


def main(sock: str) -> int:
    conn = kittytk.dial(sock, "KittyTK Demo")
    # The demo is a multi-window app: declare it so the host allows the
    # additional (normal) windows the smoke test builds below.
    conn.set_app("multiwindow")

    ui = conn.build(scripts.main_build_script())
    for key in ["w", "tabs", "binput", "wfont", "dfont", "grid",
                "bgdef", "bggreen", "bggray", "sbgdef", "sbggreen", "sbggray",
                "mdi", "mdistatus", "mdidock", "mb", "sb"]:
        if ui.id(key) == 0:
            print("FAIL surfaced id %s missing" % key, flush=True)
            return 1

    conn.build(scripts.PROTOCOL_WINDOW_SCRIPT)
    conn.exec(scripts.ABOUT_DIALOG_SCRIPT)

    child = conn.build(scripts.mdi_child_script(1))
    win_id = child.id("wwin")
    if win_id == 0:
        print("FAIL mdi child window id missing", flush=True)
        return 1
    conn.exec('set mdidock children={e1=new dockentry caption="Document 1" window=%d}' % win_id)

    tabs = ui.object("tabs")
    tabs.set("background=green")
    tabs.set('background="#333333"')
    tabs.set("background=default")

    win = ui.object("w")
    win.set('font="tuesday12"')
    win.set("denomination=32")

    for verb in ['status text="hi there"', "cut", "copy", "paste", "selectall",
                 "tile", "cascade", "theme", "desktopfont tuesday",
                 "desktopfont default", "announce_visual", "announce_speak", "rawkey"]:
        conn.exec(verb)

    print("OK", flush=True)
    conn.close()
    return 0


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("usage: demoapp_smoke.py <socket>", file=sys.stderr)
        sys.exit(64)
    sys.exit(main(sys.argv[1]))

File: c/demoapp_smoke.c
/* demoapp_smoke.c - full-demo build smoke for the C client (mirror of
 * demoapp_smoke.py). Builds the whole main window + companion window +
 * dialog + MDI document over a REAL Go host and exercises the properties
 * and every desktop-action app verb. Prints OK on success. */
#include "kittytk.h"
#include "scripts.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main(int argc, char **argv) {
    if (argc != 2) { fprintf(stderr, "usage: demoapp_smoke <socket>\n"); return 64; }
    kt_conn *c = kt_dial(argv[1], "KittyTK Demo");
    if (!c) { fprintf(stderr, "dial failed\n"); return 1; }

    char *main_src = main_build_script();
    kt_ui *ui = kt_build(c, main_src);
    free(main_src);
    if (!ui) { fprintf(stderr, "build main failed\n"); return 1; }

    const char *keys[] = {"w", "tabs", "binput", "wfont", "dfont", "grid",
                          "bgdef", "bggreen", "bggray", "sbgdef", "sbggreen", "sbggray",
                          "mdi", "mdistatus", "mdidock", "mb", "sb"};
    for (size_t i = 0; i < sizeof keys / sizeof keys[0]; i++)
        if (kt_ui_id(ui, keys[i]) == 0) { printf("FAIL surfaced id %s missing\n", keys[i]); return 1; }

    char *pw = protocol_window_script();
    if (!kt_build(c, pw)) { printf("FAIL protocol window\n"); return 1; }
    free(pw);

    char *about = about_dialog_script();
    kt_exec(c, about);
    free(about);

    char *mc = mdi_child_script(1);
    kt_ui *child = kt_build(c, mc);
    free(mc);
    if (!child || kt_ui_id(child, "wwin") == 0) { printf("FAIL mdi child\n"); return 1; }
    char dockset[128];
    snprintf(dockset, sizeof dockset,
             "set mdidock children={e1=new dockentry caption=\"Document 1\" window=%llu}",
             (unsigned long long)kt_ui_id(child, "wwin"));
    kt_exec(c, dockset);

    uint64_t tabs = kt_ui_id(ui, "tabs");
    kt_set(c, tabs, "background=green");
    kt_set(c, tabs, "background=\"#333333\"");
    kt_set(c, tabs, "background=default");

    uint64_t win = kt_ui_id(ui, "w");
    kt_set(c, win, "font=\"tuesday12\"");
    kt_set(c, win, "denomination=32");

    const char *verbs[] = {"status text=\"hi there\"", "cut", "copy", "paste", "selectall",
                          "tile", "cascade", "theme", "desktopfont tuesday",
                          "desktopfont default", "announce_visual", "announce_speak", "rawkey"};
    for (size_t i = 0; i < sizeof verbs / sizeof verbs[0]; i++)
        if (kt_exec(c, verbs[i]) != 0) { printf("FAIL app verb %s\n", verbs[i]); return 1; }

    printf("OK\n");
    fflush(stdout);
    kt_ui_free(child);
    kt_ui_free(ui);
    kt_close(c);
    return 0;
}
